src/rebar_prv_upgrade.erl

```diff
@@ -26,33 +29,105 @@ init(State) ->
                                                    {module, ?MODULE},
                                                    {bare, false},
                                                    {deps, ?DEPS},
-                                                   {example, "rebar upgrade cowboy"},
-                                                   {short_desc, "Upgrade dependency."},
-                                                   {desc, ""},
+                                                   {example, "rebar upgrade [cowboy[,ranch]]"},
+                                                   {short_desc, "Upgrade dependencies."},
+                                                   {desc, "Upgrade project dependecies. Mentioning no application "
+                                                          "will upgrade all dependencies. To upgrade specific dependencies, "
+                                                          "their names can be listed in the command."},
                                                    {opts, [
-                                                          {package, undefined, undefined, string, "Package to upgrade."}
+                                                          {package, undefined, undefined, string,
+                                                           "List of packages to upgrade. If not specified, all dependencies are upgraded."}
                                                           ]}])),
     {ok, State1}.
 
 -spec do(rebar_state:t()) -> {ok, rebar_state:t()} | {error, string()}.
 do(State) ->
     {Args, _} = rebar_state:command_parsed_args(State),
-    Name = ec_cnv:to_binary(proplists:get_value(package, Args)),
-    Locks = rebar_state:get(State, locks, []),
+    Locks = rebar_state:get(State, {locks, default}, []),
+    Deps = rebar_state:get(State, deps),
+    Names = parse_names(ec_cnv:to_binary(proplists:get_value(package, Args, <<"">>)), Locks),
+    case prepare_locks(Names, Deps, Locks, []) of
+        {error, Reason} ->
+            {error, Reason};
+        {Locks0, _Unlocks0} ->
+            Deps0 = top_level_deps(Deps, Locks),
+            State1 = rebar_state:set(State, {deps, default}, Deps0),
+            State2 = rebar_state:set(State1, {locks, default}, Locks0),
+            State3 = rebar_state:set(State2, upgrade, true),
+            Res = rebar_prv_install_deps:do(State3),
+            case Res of
+                {ok, State4} ->
+                    info_useless(
+                        [element(1,Lock) || Lock <- Locks],
+                        [rebar_app_info:name(App) || App <- rebar_state:lock(State4)]
+                    ),
+                    rebar_prv_lock:do(State4);
+                _ ->
+                    Res
+            end
+    end.
+
+-spec format_error(any()) -> iolist().
+format_error({unknown_dependency, Name}) ->
+    io_lib:format("Dependency ~ts not found", [Name]);
```

Comment:
Yeah, if we want to support unicode, ~ts tells rebar to do so. We should additionally add a call to `io:setopts([{encoding, unicode}])` in the escript-only code to make sure it works. I can reverse that bit though.